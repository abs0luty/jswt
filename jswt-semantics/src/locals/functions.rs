use super::LocalSemanticResolver;
use crate::SemanticError;
use jswt_ast::{visit::Visitor, FunctionDeclarationElement};
use jswt_common::Spannable;
use jswt_symbols::{ScopedSymbolTable, Symbol};

pub struct FunctionsLocalContext<'a> {
    symbols: &'a mut ScopedSymbolTable,
    errors: &'a mut Vec<SemanticError>,
}

impl<'a> FunctionsLocalContext<'a> {
    pub fn new(resolver: &'a mut LocalSemanticResolver) -> Self {
        Self {
            symbols: resolver.symbols,
            errors: &mut resolver.errors,
        }
    }
}

impl<'a> Visitor for FunctionsLocalContext<'a> {
    fn visit_function_declaration(&mut self, node: &FunctionDeclarationElement) {
        self.symbols.push_scope(node.body.span());
        // Redefinition errors are handled by the GlobalSemanticResolver
        // during clobal symbol resolution

        //  Add function parameters as variables in scope
        for param in node.params.parameters.iter() {
            // Check to see if the parameter is already defined
            let param_name = &param.ident.value;
            if let Some(_) = self.symbols.lookup(param_name) {
                let error = SemanticError::VariableAlreadyDefined {
                    name: param_name.clone(),
                    span: param.span(),
                };
                self.errors.push(error);
            }

            // Resolve Type from Type Annotation
            self.symbols
                .define(param_name, Symbol::ty(param.type_annotation.ty.clone()));
        }

        self.symbols.pop_scope();
    }
}

#[cfg(test)]
mod test {

    use crate::GlobalSemanticResolver;

    use super::*;
    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_symbols::BindingsTable;
    use jswt_tokenizer::Tokenizer;

    #[test]
    fn test_function_parameters_are_in_scope() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_function_parameters_are_in_scope",
            r"
        function test(x: number, y: number) {
            x;
        }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = ScopedSymbolTable::default();
        let mut bindings = BindingsTable::default();

        let mut global = GlobalSemanticResolver::new(&mut bindings, &mut symbols);
        global.resolve(&ast);
        let mut local = LocalSemanticResolver::new(&mut bindings, &mut symbols);
        local.resolve(&ast);

        assert_debug_snapshot!(local);
    }

    #[test]
    fn test_error_on_undefined_variable() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_error_on_undefined_variable",
            r"
        function test() {
            x;
        }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = ScopedSymbolTable::default();
        let mut bindings = BindingsTable::default();

        let mut global = GlobalSemanticResolver::new(&mut bindings, &mut symbols);
        global.resolve(&ast);
        let mut local = LocalSemanticResolver::new(&mut bindings, &mut symbols);
        local.resolve(&ast);

        assert_debug_snapshot!(local);
    }

    #[test]
    fn test_function_parameter_redefinition() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_function_parameter_redefinition",
            r"
        function test(x: number, x: number) {
            x;
        }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = ScopedSymbolTable::default();
        let mut bindings = BindingsTable::default();

        let mut global = GlobalSemanticResolver::new(&mut bindings, &mut symbols);
        global.resolve(&ast);
        let mut local = LocalSemanticResolver::new(&mut bindings, &mut symbols);
        local.resolve(&ast);

        assert_debug_snapshot!(local);
    }
}
