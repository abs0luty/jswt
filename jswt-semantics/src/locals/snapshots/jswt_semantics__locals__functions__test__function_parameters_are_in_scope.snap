---
source: jswt-semantics/src/locals/functions.rs
expression: local
---
LocalSemanticResolver {
    symbols: ScopedSymbolTable {
        scopes: [
            Span {
                file: "program",
                module: "root",
                start: 0,
                end: 0,
            },
        ],
        table: {
            Span {
                file: "program",
                module: "root",
                start: 0,
                end: 0,
            }: Scope {
                symbols: {
                    "test": Function(
                        FunctionSignature {
                            params: [
                                Binding(
                                    "number",
                                ),
                                Binding(
                                    "number",
                                ),
                            ],
                            returns: Binding(
                                "void",
                            ),
                        },
                    ),
                },
                ret: None,
            },
            Span {
                file: "test_function_parameters_are_in_scope",
                module: "module/test_function_parameters_are_in_scope",
                start: 45,
                end: 71,
            }: Scope {
                symbols: {
                    "x": Type(
                        TypeSignature {
                            ty: Binding(
                                "number",
                            ),
                        },
                    ),
                    "y": Type(
                        TypeSignature {
                            ty: Binding(
                                "number",
                            ),
                        },
                    ),
                },
                ret: None,
            },
        },
    },
    bindings: BindingsTable {
        bindings: {},
    },
    errors: [],
}
